### PR TITLE
Begin constructing the runtime safepoint table

### PR DESCRIPTION
**Pull request by jacob-hughes**
This generates a safepoint table which can be used at runtime by the collector. It uses `ykstackmaps` to parse the `.llvm_stackmaps` section in the Rust program's object file. The table can then be used by the collector for querying which roots are live at a safepoint. At present, only on-stack-roots have been implemented for safepoints at function calls.

This build is expected to fail until a PR is closed on `ykstackmaps` which makes some fields on `SMRec` structs public.

### CODE REVIEW

File: src/safepoints.rs
Comment (ltratt) [inline src/safepoints.rs]: Is this supposed to be a doc comment too?

File: src/safepoints.rs
Comment (ltratt) [inline src/safepoints.rs]: This feels like it should probably go down to about line 70 or so?

Also, does this location have a name that we can make a constant? In general, it's better to have named constants for this sort of thing.

File: src/safepoints.rs
Comment (ltratt) [inline src/safepoints.rs]: Let's kill this blank line and the one two lines below too.

File: src/safepoints.rs
Comment (ltratt) [inline src/safepoints.rs]: In general, you might as well use `AsRef` with `Path` so you'd have something like:

```
pub fn gen_safepoint_table<P: AsRef<Path>>(path: P) -> ...
```

because it makes the API a little easier to use in some situations.

File: src/safepoints.rs
Comment (ltratt) [inline src/safepoints.rs:128]: It feels a bit odd / delicate to have this be initialised before `parser.iter_functions`, but maybe that's just my ignorance about how it's being used.

**Comment by ltratt** [comment]
Should we be borsifying this? Maybe in a future PR? Also, we probably want to use the softdev rustfmt config (and to enforce that in Travis / bors).

**Comment by jacob-hughes** [comment]
This will need moving to bors in the future, but I think it's best to do that with the PR that introduces functionality which makes use of `rustcgc`. I'll add the softdev rustfmt config to the build though.

File: src/lib.rs
Comment (vext01) [inline src/lib.rs]: Is that required in Rust 2018? And at least one other extern site.

File: src/safepoints.rs
Comment (vext01) [inline src/safepoints.rs:4]: Why the blank line?

Comment (vext01) [reply src/safepoints.rs:4]: Also did you rustfmt this?

File: src/safepoints.rs
Comment (vext01) [inline src/safepoints.rs:20]: `is a pointer an object`

in *to*?

File: src/safepoints.rs
Comment (vext01) [inline src/safepoints.rs]: Nitpick. All your other comments have a full stop on the end.

File: src/safepoints.rs
Comment (vext01) [inline src/safepoints.rs:37]: Is performance critical here? Since the set of possible registers is fixed (and small?), you could use a bitset?

File: src/safepoints.rs
Comment (vext01) [inline src/safepoints.rs:45]: Might be worth adding a short one-liner of what this function does before the details?

File: src/safepoints.rs
Comment (vext01) [reply src/safepoints.rs]: Even if it doesn't, I think we'd benefit from a constant with a home-grown name, e.g. `NUM_SKIP_STACKMAPS`.

File: src/safepoints.rs
Comment (vext01) [inline src/safepoints.rs]: Is the scope needed here? You borrow from `stackmaps`, but only ever immutably AFAICS.

File: src/safepoints.rs
Comment (vext01) [inline src/safepoints.rs:75]: Laurie, would you recommend `try_from` here?

Jake's `as` is safe, but IIRC, a `try_from` up-cast from `u32` to `usize` would be free on x86_64.

File: src/safepoints.rs
Comment (vext01) [inline src/safepoints.rs]: `should all for GC pointers`

should all *be* for?

File: src/safepoints.rs
Comment (vext01) [inline src/safepoints.rs]: Check all of your comments for trailing stops ;)

File: src/safepoints.rs
Comment (vext01) [inline src/safepoints.rs]: I guess we'll need error handling at some time in the future.

File: src/safepoints.rs
Comment (vext01) [reply src/safepoints.rs:128]: It makes sense the way Jake has it because he's iterating over the functions and for each taking a variable number of items from `stackmaps`.

You could initialise the function iterator first, but you'd have on more line than is necessary.

File: src/safepoints.rs
Comment (vext01) [inline src/safepoints.rs:133]: This maybe changes depending on what Laurie says about using `try_from`.

**Review by vext01 — DISMISSED** [review]
Looks pretty good to me. Some comments.

File: src/safepoints.rs
Comment (jacob-hughes) [reply src/safepoints.rs]: I agree, I'll fix this

File: src/safepoints.rs
Comment (jacob-hughes) [reply src/safepoints.rs]: Ok

File: src/safepoints.rs
Comment (jacob-hughes) [reply src/safepoints.rs]: Nice, I'll do this

**Comment by vext01** [comment]
> Should we be borsifying this? 

If we do, you probably want to use binary snapshots of rustgc, like we do with ykrustc? Once I've merged my PR for that, I can show you how it's done.

File: src/safepoints.rs
Comment (jacob-hughes) [reply src/safepoints.rs:20]: Fixed

File: src/safepoints.rs
Comment (jacob-hughes) [reply src/safepoints.rs]: Okay I'll fix this for consistency

File: src/safepoints.rs
Comment (jacob-hughes) [reply src/safepoints.rs]: Good catch, yes it's unnecessary.

**Comment by jacob-hughes** [comment]
> Once I've merged my PR for that, I can show you how it's done.

Thanks!

I've pushed some fixes. I did run this through rustfmt but admittedly not using the softdev config. I think the best thing to do would be to re-format this when squashing so that it doesn't interfere with any other comments you might have.

File: src/safepoints.rs
Comment (ltratt) [reply src/safepoints.rs:128]: I guess my question is: how do we know we're taking the right number of things from `iter_stackmaps`?

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs]: Good point. Yes, we should use the 2018 edition unless there's a compelling reason not to.

File: src/safepoints.rs
Comment (ltratt) [reply src/safepoints.rs:75]: Sadly, yes, `u32` to `usize` isn't guaranteed to be safe :/ There are then two options:

1) `try_from`
2) Somewhere have an assert along the lines of `assert!(usize::max_value() > u32::max_value())`.

File: src/safepoints.rs
Comment (ltratt) [reply src/safepoints.rs]: For something like this, `unwrap` is probably OK for a long time to come: it's not as if the caller can do much else when an error is reported... There again, we could have the return type be `Result<HashMap<...>, Box<Error>>` which is a get-out-fail-freeish card in terms of percolating errors up the call stack easily.

**Comment by ltratt** [comment]
I'm fine with borsifying in another PR. I suggest we do a `rustfmt` commit now-ish (since our complaints are rapidly reducing in number), so that it doesn't appear "magically" during squashing.

File: src/safepoints.rs
Comment (vext01) [reply src/safepoints.rs:128]: Each function in the stackmap section has a `record_count` field. All we can do is trust that.

```
for sm in stackmaps.take(func.record_count() as usize) {
```

File: src/lib.rs
Comment (vext01) [reply src/lib.rs]: He already has `edition = "2018"`, so I'm surprised it didn't complain. Sure you didn't miss warnings Jake?

File: src/safepoints.rs
Comment (vext01) [reply src/safepoints.rs:75]: > u32 to usize isn't guaranteed to be safe

It is for x86_64. I was under the assumption this isn't portable code, as we are already talking about x86 registers.

File: src/safepoints.rs
Comment (vext01) [reply src/safepoints.rs]: Yeah, I pretty much default to `Box<Error>`-style errors unless the code is performance critical.

File: src/safepoints.rs
Comment (jacob-hughes) [reply src/safepoints.rs:75]: The LLVM statepoint API is only compatible with x86_64, so the cast is safe in this case. Whether we should be future-proofing it for other architectures is another question. As far as I'm aware, there are no plans yet to port the statepoint API to other architectures, although the changes would be mostly mechanical.

Comment (vext01) [reply src/safepoints.rs:75]: Why don't we use a try_from and be done with it? Presumably we won't be repeatedly loading the stackmaps section in, so it should be a small one-time cost.

Comment (ltratt) [reply src/safepoints.rs:75]: I'm fine with assuming 64 bits if a) we say so somewhere b) have an assert to enforce that. The disaster zone is that one day stackmaps are ported to (say) a 16 bit (or 128 bit) architecture and this crate silently starts doing something weird.

Comment (jacob-hughes) [reply src/safepoints.rs:75]: That's true in this case, but there will be other places in the future where these sorts of casts will be necessary and the cost would then be incurred on each collection. In addition, a lot of the stack walking code is platform specific and not portable.

Comment (jacob-hughes) [reply src/safepoints.rs:75]: @ltratt A disclaimer seems like a good idea .

File: src/lib.rs
Comment (jacob-hughes) [reply src/lib.rs]: Hmm, no warnings but I've removed the duplicate extern site as that shouldn't have been there.

**Comment by jacob-hughes** [comment]
Think I've addressed all the comments, let me know if you're happy for me to squash.

File: src/safepoints.rs
Comment (ltratt) [reply src/safepoints.rs:75]: I really think we should have an `assert` or similar in that checks this somewhere that's guaranteed to be called at least once so that if someone does run it on an inappropriate machine it goes splat. 

A possibly good approach is to use https://crates.io/crates/static_assertions -- `assert_eq_size` might be the right thing for us.

File: src/safepoints.rs
Comment (ltratt) [reply src/safepoints.rs:128]: OK.

File: src/safepoints.rs
Comment (jacob-hughes) [reply src/safepoints.rs:75]: Ah, sorry I missed this one. I'll add an assertion at the beginning of the table construction function.

Comment (ltratt) [reply src/safepoints.rs:75]: OK.

**Comment by jacob-hughes** [comment]
Added a static assertion to check that we use 64bit word sizes

File: Cargo.toml
Comment (ltratt) [inline Cargo.toml]: Should probably just be `"0.3"` so we pick up point releases automatically.

File: src/lib.rs
Comment (ltratt) [inline src/lib.rs]: I don't think we need this line? Or the one below?

**Comment by jacob-hughes** [comment]
Fixed

**Comment by ltratt** [comment]
Please squash.

**Comment by jacob-hughes** [comment]
Squashed

**Comment by ltratt** [comment]
I think we can do some more squashing here: https://github.com/softdevteam/rgcrt/pull/1/commits/b35d20752ae0cfa028566ad274404411f22c6b32 and https://github.com/softdevteam/rgcrt/pull/1/commits/92f8cac1169149760a43c04a5f146a9cde041161 should probably go together; and maybe https://github.com/softdevteam/rgcrt/pull/1/commits/e33b893d7f91db335f49de0f17e7b25652455668 can be squashed in elsewhere?

**Comment by jacob-hughes** [comment]
Sure. What about now?

**Comment by ltratt** [comment]
Works for me. I'll merge when the tests have passed.

For the next PR, whether you use Travis or bors, can you make sure that whichever of them enforces rustfmt? e.g. here's a Travis example https://github.com/softdevteam/grmtools/blob/master/.travis.yml

**Pull request merged**
